fvm/src/syscalls/mod.rs

```diff
@@ -87,32 +94,47 @@ pub fn charge_for_exec<K: Kernel>(
     let mut ctx = ctx.as_context_mut();
     let global = ctx.data_mut().avail_gas_global;
 
-    let milligas_available = global
+    let gas_available = global
         .get(&mut ctx)
         .i64()
         .context("failed to get wasm gas")
-        .map_err(Abort::Fatal)?;
-
-    // Determine milligas used, and update the gas tracker.
-    let mut exec_gas = {
+        .map_err(Abort::Fatal)?
+        // -> u64
+        .try_into()
+        .map(Gas::new)
+        // The gas goes below zero when we run out, so we set "gas available" to 0.
+        .unwrap_or_default();
+
+    // Determine gas used, and update the gas tracker.
+    let mut exec_gas_charge = {
         let data = ctx.data_mut();
-        let last_milligas = mem::replace(&mut data.last_milligas_available, milligas_available);
-        // This should never be negative, but we might as well check.
-        Gas::from_milligas(last_milligas.saturating_sub(milligas_available))
+        let prev_gas_available = mem::replace(&mut data.last_gas_available, gas_available);
+        prev_gas_available - gas_available
     };
 
     let data = ctx.data_mut();
 
     // Separate the amount of gas charged for memory; this is only makes a difference in tracing.
     let memory_bytes = data.kernel.limiter_mut().memory_used();
     let memory_delta_bytes = memory_bytes.saturating_sub(data.last_memory_bytes);
-    let memory_gas = data.kernel.price_list().grow_memory_gas(memory_delta_bytes);
 
-    exec_gas = (exec_gas - memory_gas).max(Gas::zero());
+    // `exec_gas_charge` is the number we want to charge. If, for some reason, `memory_gas_charge`
+    // exceeds `exec_gas_charge`, we just set `memory_gas_charge` to `exec_gas_charge`, and set
+    // `exec_gas_charge` to zero.
+    let mut memory_gas_charge = data.kernel.price_list().grow_memory_gas(memory_delta_bytes);
+    if memory_gas_charge <= exec_gas_charge {
+        exec_gas_charge -= memory_gas_charge;
+    } else {
+        memory_gas_charge = exec_gas_charge;
+        exec_gas_charge = Gas::zero();
+    }
```

Comment:
I don't think this could have ever been an issue, but... we might as well be careful.